pkg/scalers/scaler.go

```diff
@@ -168,47 +168,106 @@ func GenerateMetricInMili(metricName string, value float64) external_metrics.Ext
 	}
 }
 
+type Option func(*configOptions)
+
+type configOptions struct {
+	parameter         string
+	useMetadata       bool
+	useAuthentication bool
+	useResolvedEnv    bool
+	isOptional        bool
+	defaultVal        interface{}
+	targetType        reflect.Type
+}
+
+func WithParameter(parameter string) Option {
+	return func(opt *configOptions) {
+		opt.parameter = parameter
+	}
+}
+
+func UseMetadata(metadata bool) Option {
+	return func(opt *configOptions) {
+		opt.useMetadata = metadata
+	}
+}
+
+func UseAuthentication(auth bool) Option {
+	return func(opt *configOptions) {
+		opt.useAuthentication = auth
+	}
+}
+
+func UseResolvedEnv(resolvedEnv bool) Option {
+	return func(opt *configOptions) {
+		opt.useResolvedEnv = resolvedEnv
+	}
+}
+
+func IsOptional(optional bool) Option {
+	return func(opt *configOptions) {
+		opt.isOptional = optional
+	}
+}
+
+func WithDefaultVal(defaultVal interface{}) Option {
+	return func(opt *configOptions) {
+		opt.defaultVal = defaultVal
+	}
+}
+
+func WithTargetType(targetType reflect.Type) Option {
```

Comment:
I think the type shouldn't be an option, ideally, we can further improve this with generics and infer the type from invocations also guarded by proper type constraint so users don't get hit by runtime errors from `convertToType` type switch.

Reply:
Thanks for the swift review. Agree with you those two args are compulsory so removed `WithTargetType` and `WithParameter` . I did tried to incorporate some changes from  your PR to use generics type but still have not get it to pass the unittest. Generics is a (fairly) new feature for golang so I'm learning something new today :smile: . If you have time, please help me to incorporate it (It can be a separate PR after this one is merged for your ease of development)